### ISSUE TITLE
current folder not getting resolved on linux 

### ISSUE THREAD

**Issue by zeusintuivo** [wez/wezterm#418]
## Describe the bug

When control clicking a file, it does not find the current folder. 

## Environment (please complete the following information):

 - OS: Gnome, Linux Fedora 32
 - Version: wezterm 20201101-103216-403d002d

## To Reproduce

First add the hyperfile config to make open files https://github.com/wez/wezterm/pull/412#issuecomment-753428584 

Go to a folder with a project outside of home like `/home/user/project `
then run the search command  `grep -Rn hello index.html`
Should see something like 
`index.html:10`
Control click filename and should open 
 

## Configuration

Using the config like stated here https://github.com/wez/wezterm/pull/412#issuecomment-753428584 

## Expected behavior

Control click should open file 

## Screenshots

NA 

## Session Recording

NA

## Additional context

Inside Mac wezterm resolved the current directory but in Linux it attempts open the file from $HOME. 
so if the terminal is currently at /home/user/project/  and search finds `index.html:10`  
it should open it, but in Lunix it sends `/home/user/index.html:10` when it should be  `/home/user/project/index.html:10`

**Comment by wez** [comment]
Have you configured terminal/shell integration in your shell?

There's some information in the docs here: https://wezfurlong.org/wezterm/shell-integration.html

This shell file configures your shell to emit OSC 7 and OSC 133 in the appropriate places:
https://github.com/wez/wezterm/blob/master/assets/shell-integration/wezterm.sh

(that file is installed by the nightly fedora rpms such that bash and zsh will source it automatically)

Without the shell integration, wezterm can try to guess the cwd by snooping the pty, but it is not guaranteed to produce correct results.  I'm assuming that something like that is happening for you, as I did try exactly the scenario you described when I tried out the config initially and it worked for me in that case!

**Comment by zeusintuivo** [comment]
I do not understand why is design of wezterm so 'hacky' 

Is like so very customizable that everything inside wezterm needs to be coded and researched before starting to use a half way decent terminal. When using iTerm2 there is no need to "hack" it when starting to use it. Is like some very basic functionalities should be available by default. 

I thought wezterm should have lots of potential because it uses Rust, but perhaps is far away from it. 

Regarding this issue, all these terms "OSC 7 " and "OSC 133" do not mean anything to me. Perhaps they are important, but since I am not around this world. It makes no sense to mention them. Perhaps around the "terminal hackers" 

In conclusion seems, that the stability of wezterm, depends of my patience and ability to understand the wezterm docs, and apply  them correctly.

**Comment by wez** [comment]
I appreciate that terminals are hard to understand: they require some low level systems programming that places a number of constraints on what they can do.  Supporting three different operating system families makes this more of a challenge.

I don't think it is appropriate for you to decide that the design is 'hacky' when you're not familiar with how terminals work and haven't looked at the implementation of other terminals in order to make an informed comparison.

It is fair to identify that the out-of-the-box experience isn't as polished as software that has been around for over a decade, but you don't have to be insulting about it.

I'd prefer it if you could focus the dialog in these issues on productively identifying problems so that we can get closer to a more magical out-of-the-box experience.

I have put a lot of time into the documentation, and that tries to explain OSC 7 and OSC 133 and the associated problems that they aim to solve.  If you read it and don't understand it, I'd appreciate feedback on where it stopped making sense to you.  If you didn't read it and don't want to read it, I understand, but humbly suggest that wezterm may not be for you at this time.

Please also understand that I'm not paid to work on this project, and you're getting it for free with no warranty for any purpose.
I do this for fun and I'm much more willing to spend my limited time working through things with you if you keep things respectful and productive.

**Comment by bew** [comment]
I think that one of the misunderstanding is what a terminal emulator actually is, what role does it plays, and the difference between the terminal emulator and the shell.

@zeusintuivo, as you said you had no idea what Wez was talking about, I suggest you to read these explanations on the topic:
https://askubuntu.com/questions/506510/what-is-the-difference-between-terminal-console-shell-and-command-line
this one is shorter but has less details
https://superuser.com/questions/880344/what-is-the-difference-between-terminal-and-bash

As for this issue, keep in mind that the terminal is only showing to you what the shell tells it to show. When you type `cd project` then `Enter`, here is what happens (without going too much in details):
Each key you press is send to the terminal, which sends them to the shell program, when the shell receives the command and the `Enter` key, it will change it's current working directory to `project`, and give you the prompt again so you can type another command.
The terminal has NO knowledge of the current directory change of the shell, unless the shell tells the terminal that its current directory changed.
The way the shell can communicate with the terminal is by trying to display some sort of special characters. Since the terminal is the one responsible to actually show to you the user the characters, it receives those special characters, and knows that they mean something specific and should not be displayed.

There are many of these special characters, used to control the terminal (move the cursor, change text color, clear text, etc...), some are standard and some are not. One of them is OSC 7 that is recognized by some terminal as a way to tell it that the current working directory changed.

I hope this made a little bit more sense!

---
@wez maybe an explanation like this can be added in your docs?
I wrote this quickly, I'm sure you'll do a better write up on the topic since you worked on it for so long :smiley:

**Comment by zeusintuivo** [comment]
> I appreciate that terminals are hard to understand: they require some low level systems programming that places a number of constraints on what they can do. Supporting three different operating system families makes this more of a challenge.
> 
> I don't think it is appropriate for you to decide that the design is 'hacky' when you're not familiar with how terminals work and haven't looked at the implementation of other terminals in order to make an informed comparison.
> 
> It is fair to identify that the out-of-the-box experience isn't as polished as software that has been around for over a decade, but you don't have to be insulting about it.
> 
> I'd prefer it if you could focus the dialog in these issues on productively identifying problems so that we can get closer to a more magical out-of-the-box experience.
> 
> I have put a lot of time into the documentation, and that tries to explain OSC 7 and OSC 133 and the associated problems that they aim to solve. If you read it and don't understand it, I'd appreciate feedback on where it stopped making sense to you. If you didn't read it and don't want to read it, I understand, but humbly suggest that wezterm may not be for you at this time.
> 
> Please also understand that I'm not paid to work on this project, and you're getting it for free with no warranty for any purpose.
> I do this for fun and I'm much more willing to spend my limited time working through things with you if you keep things respectful and productive.

I appreciate that this is open source. That is low cheap and insulting statement from your part. I am only discussing the part that why do, as the owner the project, decide the make it a hacky software.  So, my time reviewing software is as important as your time doing it, and you also get my reviews for free. 

Not just that, I do not appreciate that you punish my effort and interest in your project this way. I did not decide how to make your project. 

About other terminals, I have made a study on all several terminals that I could find and get my hands on. I looked into wezterm, because it looked promising. But it seems, like deep whole of knowledge just to make work with simple functions. 

As a developer, I look for something that can reliable. If the instructions manual, is so complex, like a its operating system. It seems like the terminal is more than terminal and more like operating system that needs to be configured before stating to use. With unexpected results for me so far. 

As for better terminals, I can only speak for iTerm in Mac or even for open source: Alacritty. I did try about 10 more of all the others. Kitty, Upterm, etc etc, even Edex-UI. 

Regarding other Linux terminals, there are several open source, like Guake on the top of the list. Others like Terminator or Tilix , etc etc, are just not as good. Guake is far superior and also open source. 

I was hoping that wezterm could be a runner up regarding a good terminal. But it seems that for every little thing, there is a lenghty documentation and extra steps that need to taken just to accomplish little thing working. 

On Mac, I ask myself the following question? On Mac is there an open source terminal that can be used instead of the default Terminal.app or iTerm.app that can be used and with a better usage experience ?

On Linux, I ask myself the following question? On Linux, is there an open source terminal that can be used instead of the default Gnome-Terminal or the awesome Guake  that can be used and with an alternative customizable experience ?

On Windows, I ask myself the following question? On Windows, is there an open source terminal that can be used instead of the CMD Terminal or Powershell defaults that can be used with closer experience to a Linux software ?

**Comment by zeusintuivo** [comment]
> The terminal has NO knowledge of the current directory change of the shell, unless the shell tells the terminal that its current directory changed.



I don't buy the folder explanation. Take a look at Guake's Quick Open File functionality, how do they accomplish it ?

**Comment by wez** [comment]
> I appreciate that this is open source. That is low cheap and insulting statement from your part. I am only discussing the part that why do, as the owner the project, decide the make it a hacky software. So, my time reviewing software is as important as your time doing it, and you also get my reviews for free.
> Not just that, I do not appreciate that you punish my effort and interest in your project this way. I did not decide how to make your project.

I'm not sure how my request for you to be a productive and respectful member of the wezterm community came across as punitive or insulting.  I do expect members of the community to "be excellent to each other" and I will comment when that doesn't appear to be the case.

I recognize that you are frustrated and that you are new to understanding the technical details that apply to this project, and that English may not be your first language, and that it may not be your intent to do so, but even so: you are repeatedly coming across as impatient, rude, insulting and argumentative.

Nobody enjoys being told something like that, and it can trigger the fight/flight response and that can in turn lead to additional or escalating "heat" in these sorts of discussions.

I encourage you to take the time to read through this thread again, internalize it, and to **wait until you are in a calm state before responding again**.  It takes time and effort to find phrasing that avoids escalating a debate.  I find that it is generally best to keep dialog focused on the technical/factual aspects to avoid conflict.

Sharing your opinion about how the software should behave is welcomed, but repeated statements about your opinion of a more abstract concept, such as overall design, in the absence of a concrete suggestion for improvement, isn't productive because there isn't an actionable element.

For example: "I think that the way X works is a bit of a hack, have you considered doing Y instead?" is the start of a constructive dialog, whereas "I think you've made a hacky software" is not because there isn't a place to start working through and unwinding what the real concern is, and that sentence overall is disparaging the result of someone else's hard work.  Instead, I'd recommend identifying the component(s) and saying "I don't understand X, could you give me some more information on why it works this way?".  Then, once you have the understanding, you may be in a position to discuss alternative implementations.

**Comment by wez** [comment]
> @wez maybe an explanation like this can be added in your docs?
> I wrote this quickly, I'm sure you'll do a better write up on the topic since you worked on it for so long

Yeah, I think it would be good to have a section on "What is a Terminal" that has an architectural diagram showing the emulator, the kernel, the shell, and the processes spawned by the shell, explaining how the pty is a dumb channel for input/output, how multiple processes can be attached to a pty, how unix concepts like foreground/background, session and process groups make it difficult to know what "the" working directory is, how Windows doesn't even have those concepts, and how panes and tabs have implied context, and how multiplexers, ssh and serial connections are not local and how that implied context is unavailable, and how OSC 7 and OSC 133 add semantic meaning to the output so that the terminal emulator can improve the user experience.

If you have the time to start throwing a page like that together, @bew, I'd appreciate it; it doesn't have to be complete by any means, just getting a starting point would be helpful and we can evolve it from there.

**Comment by bew** [comment]
> > The terminal has NO knowledge of the current directory change of the shell, unless the shell tells the terminal that its current directory changed.
> 
> I don't buy the folder explanation. Take a look at Guake's Quick Open File functionality, how do they accomplish it ?

Here is how they accomplish this:
https://github.com/Guake/guake/blob/d2b8d8ef44297d3ecc0440c3d91d966499b3dc9c/guake/terminal.py#L246-L255
Basically they read the current working directory of the immediate child process of the terminal and hope that that's what the user wants.

**it works for basic cases**
It may work is most basic cases, but if they don't have any other way to get the current working directory (and I don't see any with a quick look at their code), it is a **very fragile** way to get the working directory.
It works only if you have 1 shell open in the terminal and you are actively using that shell.
Meaning that the processes are `terminal -> shell` (the shell is running as a child of the terminal).

**it breaks for anything more complex**
As soon as you start doing more complex things like running containers and getting access to them, attaching to a tmux session, controlling a remote server through ssh, or simply running another shell in your starting shell (to build a program in a restricted environment for example), the shell you're actively using is NOT the immediate child of the terminal and the trick used by Guake won't work anymore or worse, will give you wrong results.

**example**
The processes when connecting to another server would be for example:
`terminal -> shell1 -> ssh client` then, on the remote server `ssh daemon -> shell2` (and if you attach to an existing tmux session on that server you can add `tmux -> shell3`; this is what I have every day when I work, it's pretty common when working remotely I think).
Here, Guake would only know the working directory of `shell1`, even if you're actively using `shell2` or `shell3`, and when you're changing directory in those shells, the parent processes are not aware of the directory changes.

**why OSC 7 solves the problem**
Any process in the hierarchy under the terminal can usually display something, and that works because in this case, each process relays the input & output of it's immediate subprocess to its parent process, until the terminal gets it and do something useful like printing `foobar` at the cursor location.
Since any process can display something, if a special sequence of characters like the one for OSC 7 is sent, the terminal will be aware of it and can do something about it.
So if you're working in `shell3`, and configured it to send OSC 7 (with the info you want) each time you change directories, the terminal will get the information and know about the directory change.
Problem solved for simple & complex cases

---
> If you have the time to start throwing a page like that together, @bew, I'd appreciate it; it doesn't have to be complete by any means, just getting a starting point would be helpful and we can evolve it from there.

I keep that in mind :smiley: probably not this week-end though as I've already used a good part of my free time writing these :eyes:

**Comment by wez** [comment]
There's now a nice flow chart illustrating how the current working directory is resolved in the docs:
https://wezfurlong.org/wezterm/config/lua/config/default_cwd.html

There are links to that page from the [Launching Programs](https://wezfurlong.org/wezterm/config/launch.html) docs.

**Comment by kovidgoyal** [comment]
Just FYI, OSC 7 is a totally broken and unnecessary hack. 

1) You can always get the cwd of the current foreground process associated with the tty device on Unix. For code to do that see kitty or alacritty which copied it from kitty.

2) Getting the current directory over ssh is fairly meaningless, since you cant open a tab/window in that directory anyway, robustly.

3) Windows is probably the only place where OSC 7 is needed

For more detailed discussion: https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/20
